src/listener.ts

```diff
@@ -0,0 +1,145 @@
+import http from 'http'
+import { logger } from '@libp2p/logger'
+
+// @ts-expect-error no types
+import isNode from 'detect-node'
+// @ts-expect-error no types
+import wrtc from 'wrtc'
+// @ts-expect-error no types
+import SimplePeer from 'libp2p-webrtc-peer'
+import { base58btc } from 'multiformats/bases/base58'
+import { toString } from 'uint8arrays/to-string'
+import { fromString } from 'uint8arrays/from-string'
+import type { Multiaddr } from '@multiformats/multiaddr'
+import type { IncomingMessage, ServerResponse } from 'http'
+import { EventEmitter, CustomEvent } from '@libp2p/interfaces'
+import type { Connection } from '@libp2p/interfaces/connection'
+import type { Upgrader, Listener } from '@libp2p/interfaces/transport'
+import ipPortToMultiaddr from 'libp2p-utils/src/ip-port-to-multiaddr'
+
+import { toMultiaddrConnection } from './socket-to-conn'
+import type { WebRTCDirectListenerOptions } from './index'
+import type { ExtendedMultiaddrConnection } from './socket-to-conn'
+
+
+const log = logger('libp2p:webrtcdirect:listener')
+
+interface WebRTCDirectListener extends Listener {
+  __connections: ExtendedMultiaddrConnection[]
+}
+
+export function createListener (upgrader: Upgrader, options?: WebRTCDirectListenerOptions) {
+  const handler = options?.handler
+  const channelOptions = options?.channelOptions
+  const server = http.createServer()
+
+  let maSelf: Multiaddr
+
+  server.on('request', async (req: IncomingMessage, res: ServerResponse) => {
+    res.setHeader('Content-Type', 'text/plain')
+    res.setHeader('Access-Control-Allow-Origin', '*')
+
+    const path = req.url
+    const incSignalStr = path.split('?signal=')[1]
+    const incSignalBuf = base58btc.decode(incSignalStr)
+    const incSignal = JSON.parse(toString(incSignalBuf))
+
+    const channel = new SimplePeer({
+      trickle: false,
+      wrtc: isNode ? wrtc : undefined,
+      ...channelOptions
+    })
+
+    const maConn = toMultiaddrConnection(channel, {
+      remoteAddr: ipPortToMultiaddr(req.socket.remoteAddress, req.socket.remotePort)
```

Comment:
Should we throw when `req.socket.remoteAddress` is undefined? It can be, according to the docs https://nodejs.org/api/net.html#class-netsocket